tests/envs/MaMuJoCo/test_MaMuJoCo.py

```diff
@@ -0,0 +1,495 @@
+    scenario_conf("manyagent_swimmer", "6x1"),
+    scenario_conf("manyagent_ant", "2x3"),
+    scenario_conf("manyagent_ant", "3x1"),
+    scenario_conf("coupled_half_cheetah", "1p1"),
+    scenario_conf("coupled_half_cheetah", None),
+]
+
+observation_depths = [None, 0, 1, 2]
+
+
+def assert_dict_numpy_are_equal(
+    dict_a: dict[any, numpy.array], dict_b: dict[any, numpy.array]
+) -> None:
+    assert dict_a.keys() == dict_b.keys()
+    for key in dict_a.keys():
+        assert (dict_a[key] == dict_b[key]).all()
+
+
+@pytest.mark.parametrize("observation_depth", observation_depths)
+@pytest.mark.parametrize("task", pre_defined_factorizations + sample_configurations)
+def test_general(observation_depth, task):
+    parallel_api_test(
+        MultiAgentMujocoEnv(task.scenario, task.conf, agent_obsk=observation_depth),
+        num_cycles=1_000_000,
+    )
+
+
+@pytest.mark.parametrize("observation_depth", observation_depths)
+@pytest.mark.parametrize("task", pre_defined_factorizations)
+def test_action_and_observation_mapping(observation_depth, task):
+    """
+    assert that converting local <-> global <-> local obervations/actions results in the same observation/actions
+    """
+    test_env = MultiAgentMujocoEnv(
+        task.scenario, task.conf, agent_obsk=observation_depth
+    )
+    # assert action mapping
+    global_action = test_env.gym_env.action_space.sample()
+    assert (
+        global_action
+        == test_env.map_local_actions_to_global_action(
+            test_env.map_global_action_to_local_actions(global_action)
+        )
+    ).all()
+
+    if task == scenario_conf("Reacher", "2x1"):
+        return  # observation mapping not implemented on 'Reacher' Environment
+
+    # assert observation mapping
+    test_env.reset()
+    global_observations = test_env.state()
+    local_observations = test_env.unwrapped._get_obs()
+    test_env.reset()
+    assert_dict_numpy_are_equal(
+        test_env.map_global_state_to_local_observations(global_observations),
+        local_observations,
+    )
+
+
+@pytest.mark.parametrize("observation_depth", observation_depths)
+@pytest.mark.parametrize("task", sample_configurations)
+def test_action_mapping(observation_depth, task):
+    # observation mapping not implemented non-Gymansium mujoco environments
+    """
+    assert that converting local <-> global <-> local actions results in the same actions
+    """
+    test_env = MultiAgentMujocoEnv(
+        task.scenario, task.conf, agent_obsk=observation_depth
+    )
+    global_action = test_env.gym_env.action_space.sample()
+    assert (
+        global_action
+        == test_env.map_local_actions_to_global_action(
+            test_env.map_global_action_to_local_actions(global_action)
+        )
+    ).all()
+
+
+def test_k_dict():
+    """
+    asserts that obsk.get_joints_at_kdist() generates the correct observation factorization
+    The outputs have been hand written
+    """
+    scenario = "Ant"
```

Comment:
I realise that this function is much more complex, is it possible to use parameterize, if not don't worry

Reply:
Could you change to a parameterize with `"scenario, agent_conf, obsk_0_k_dicts, obsk_1_k_dicts, obsk_2_k_dicts"`

Reply:
I could do that, but that would result in 1000+ characters lines (would need to use`# noqa: F401`) are you ok with that?

Reply:
If that is true how does the PR currently pass pre-commit? 
This should be possible, I agree that it will look ugly but it will be easier to check that tests as you just look at a single test (which is repeated) rather than long list of tests